File: Week1_Design Patterns and Principles/Ex_2_Implementing_the_Factory_Method_Pattern/Code/com/example/factory/documents/WordDocument.java
package com.example.factory.documents;
import java.time.LocalDateTime;
import java.time.format.DateTimeFormatter;
public class WordDocument implements Document {
    private LocalDateTime creationTime;
    private String fileName;
    public WordDocument() {
        this.creationTime = LocalDateTime.now();
        this.fileName = "Document_" + creationTime.format(DateTimeFormatter.ofPattern("yyyyMMdd_HHmmss")) + ".docx";
    }
    @Override
    public void open() {
        System.out.println("Opening Word document: " + fileName);
        System.out.println("Loading Microsoft Word application...");
    }
    @Override
    public void save() {
        System.out.println("Saving Word document: " + fileName);
        System.out.println("Document saved in .docx format");
    }
    @Override
    public void close() {
        System.out.println("Closing Word document: " + fileName);
        System.out.println("Microsoft Word application closed");
    }
    @Override
    public void print() {
        System.out.println("Printing Word document: " + fileName);
        System.out.println("Sending to default printer...");
    }
    @Override
    public String getDocumentType() {
        return "Microsoft Word Document";
    }
    @Override
    public LocalDateTime getCreationTime() {
        return creationTime;
    }
}

File: Week1_Design Patterns and Principles/Ex_2_Implementing_the_Factory_Method_Pattern/Code/Documents/WordDocument.cs
using System;
using System.Globalization;

namespace FactoryDemo.Documents;

public sealed class WordDocument : IDocument
{
    private readonly DateTime _creationTime = DateTime.Now;
    private readonly string _fileName;

    public WordDocument()
    {
        _fileName = $"Document_{_creationTime.ToString("yyyyMMdd_HHmmss", CultureInfo.InvariantCulture)}.docx";
    }

    public void Open()  => Console.WriteLine($"Opening Word document: {_fileName}\nLoading Microsoft Word application…");
    public void Save()  => Console.WriteLine($"Saving Word document: {_fileName}\nDocument saved in .docx format");
    public void Close() => Console.WriteLine($"Closing Word document: {_fileName}\nMicrosoft Word application closed");
    public void Print() => Console.WriteLine($"Printing Word document: {_fileName}\nSending to default printer…");

    public string GetDocumentType()  => "Microsoft Word Document";
    public DateTime GetCreationTime() => _creationTime;
}
